src/lseq/mod.rs

```diff
@@ -74,121 +94,152 @@ pub enum Op<T> {
+            self.seq.insert(res, Entry { id: ix, dot, c });
         }
     }
 
     /// Remove an identifier from the LSEQ
     pub fn delete(&mut self, ix: Identifier) {
         // Deletes only have an effect if the identifier is already in the tree
-        if let Ok(i) = self.seq.binary_search_by(|e| e.0.cmp(&ix)) {
+        if let Ok(i) = self.seq.binary_search_by(|e| e.id.cmp(&ix)) {
             self.seq.remove(i);
         }
     }
 
-
-
     /// Perform a local insertion of an element at a given position.
     /// If `ix` is greater than the length of the LSeq then it is appended to the end.
-    pub fn insert_index(&mut self, ix: usize, c: T) -> Op<T> {
+    pub fn insert_index(&mut self, ix: usize, c: T) -> Op<T>
+    where
+        T: Clone,
+    {
         let lower = self.gen.lower();
         let upper = self.gen.upper();
 
         // If we're inserting past the length of the LSEQ then it's the same as appending.
         let ix_ident = if self.seq.len() <= ix {
-            let prev = self.seq.last().map(|(i, _, _)| i).unwrap_or_else(|| &lower);
+            let prev = self
+                .seq
+                .last()
+                .map(|Entry { id, .. }| id)
+                .unwrap_or_else(|| &lower);
             self.gen.alloc(prev, &upper)
         } else {
             // To insert an element at position ix, we want it to appear in the sequence between
             // ix - 1 and ix. To do this, retreive each bound defaulting to the lower and upper
             // bounds respectively if they are not found.
             let prev = match ix.checked_sub(1) {
-                Some(i) => &self.seq.get(i).unwrap().0,
+                Some(i) => &self.seq.get(i).unwrap().id,
                 None => &lower,
             };
-            let next = self.seq.get(ix).map(|(i, _, _)| i).unwrap_or(&upper);
+            let next = self
+                .seq
+                .get(ix)
+                .map(|Entry { id, .. }| id)
+                .unwrap_or(&upper);
             let a = self.gen.alloc(&prev, next);
 
             assert!(prev < &a);
             assert!(&a < next);
             a
         };
-        let op = Op::Insert{ id: ix_ident, dot: self.dot.clone(), c };
+        let op = Op::Insert {
+            id: ix_ident,
+            dot: self.dot.clone(),
+            c,
+        };
         self.dot.counter += 1;
         self.apply(op.clone());
         op
-
-
     }
 
-    /// Perform a local deletion at ix. If ix does not exist then it will delete the last element
-    /// of the tree.
-    pub fn delete_index(&mut self, mut ix: usize) -> Op<T> {
-        if ix >= self.seq.len()  {
-            ix = self.seq.len() - 1;
+    /// Perform a local deletion at ix.
+    ///
+    /// If `ix` is out of bounds, i.e. `ix > self.len()`, then
+    /// the `Op` is not performed and `None` is returned.
+    pub fn delete_index(&mut self, ix: usize) -> Option<Op<T>>
```

Comment:
I think this is a safer option. If the user _means_ to always delete the end element, they can:
a) Calculate the `len` in the case of `None`
b) Pre-calculate the `ix`

In either case, they're forced to know what to do. But at least there's option of not deleting anything in case they got the index wrong.

Reply:
ref: https://github.com/xldenis/rust-crdt/pull/1#discussion_r407458485